lerobot/scripts/train.py

```diff
@@ -420,13 +429,188 @@ def evaluate_and_checkpoint_if_needed(step):
+            "When online training is resumed we use the latest online buffer, which is not the necessarily "
+            "the one that was present at the time of the provided checkpoint!"
+        )
+    online_dataset = OnlineBuffer(
+        online_buffer_path,
+        data_spec={
+            **{k: {"shape": v, "dtype": np.dtype("float32")} for k, v in policy.config.input_shapes.items()},
+            **{k: {"shape": v, "dtype": np.dtype("float32")} for k, v in policy.config.output_shapes.items()},
+            "next.reward": {"shape": (), "dtype": np.dtype("float32")},
+            "next.done": {"shape": (), "dtype": np.dtype("?")},
+            "next.success": {"shape": (), "dtype": np.dtype("?")},
+        },
+        buffer_capacity=cfg.training.online_buffer_capacity,
+        fps=online_env.unwrapped.metadata["render_fps"],
+        delta_timestamps=cfg.training.delta_timestamps,
+    )
+
+    # If we are doing online rollouts asynchronously, deepcopy the policy to use for online rollouts (this
+    # makes it possible to do online rollouts in parallel with training updates).
+    online_rollout_policy = deepcopy(policy) if cfg.training.do_online_rollout_async else policy
+
+    # Create dataloader for online training.
+    concat_dataset = torch.utils.data.ConcatDataset([offline_dataset, online_dataset])
+    if len(offline_dataset) > 0 and len(online_dataset) > 0:
+        weights = torch.tensor(
+            [(1 - cfg.training.online_sampling_ratio) / len(offline_dataset)] * len(offline_dataset)
+            + [cfg.training.online_sampling_ratio / len(online_dataset)] * len(online_dataset)
+        )
+    elif len(offline_dataset) > 0:
+        weights = torch.ones(len(offline_dataset))
+    elif len(online_dataset) > 0:
+        weights = torch.ones(len(online_dataset))
+    sampler = torch.utils.data.WeightedRandomSampler(
+        weights, num_samples=len(concat_dataset), replacement=True
+    )
+    if cfg.training.do_online_rollout_async and cfg.training.num_workers > 0:
+        logging.warning(
+            f"{cfg.training.num_workers=} will be ignored as you have "
+            f"{cfg.training.do_online_rollout_async=}. With > 0 workers, batches are prefetched, but "
+            "this isn't set up to work with asynchronous updates to the online buffer. Instead, 0 "
+            "workers will be used for the online training dataloader."
+        )
+    dataloader = torch.utils.data.DataLoader(
+        concat_dataset,
+        batch_size=cfg.training.batch_size,
+        num_workers=cfg.training.num_workers if not cfg.training.do_online_rollout_async else 0,
+        sampler=sampler,
+        pin_memory=device.type != "cpu",
+        drop_last=True,
+    )
+    dl_iter = cycle(dataloader)
+
+    # Lock and thread pool executor for asynchronous online rollouts. When asynchronous mode is disabled,
+    # these are still used but effectively do nothing.
+    lock = Lock()
+    executor = ThreadPoolExecutor(max_workers=1)
+
```

Comment:
Impressive implementation. How slow sync is versus async? Could be worth adding an info here.

Why is max_workers set to 1?

Reply:
Regarding your question about `max_workers`, I hope this comment helps:

```
# Note: 1 worker because we only ever want to run one set of online rollouts at a time.
# Batch parallelization of rollouts is handled within the job.
```